File: web/src/api/watchList.ts
import apiClient from './index';

export type MarketType = 'CN' | 'US' | 'HK';

export interface WatchListItem {
  id: number;
  code: string;
  name: string | null;
  notes: string | null;
  market_type: MarketType;
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface WatchListResponse {
  items: WatchListItem[];
  total: number;
}

export interface WatchListItemCreate {
  code: string;
  name?: string;
  notes?: string;
  market_type?: MarketType;
  is_favorite?: boolean;
}

export interface WatchListItemUpdate {
  name?: string;
  notes?: string;
  market_type?: MarketType;
  is_favorite?: boolean;
}

export const watchListApi = {
  async list(): Promise<WatchListResponse> {
    const res = await apiClient.get('/api/v1/watch-list');
    return res.data as WatchListResponse;
  },

  async create(body: WatchListItemCreate): Promise<WatchListItem> {
    const res = await apiClient.post('/api/v1/watch-list', body);
    return res.data as WatchListItem;
  },

  async update(id: number, body: WatchListItemUpdate): Promise<WatchListItem> {
    const res = await apiClient.put(`/api/v1/watch-list/${id}`, body);
    return res.data as WatchListItem;
  },

  async remove(id: number): Promise<void> {
    await apiClient.delete(`/api/v1/watch-list/${id}`);
  },
};

File: api/v1/endpoints/watch_list.py
# -*- coding: utf-8 -*-
"""
===================================
自选股接口 (watch_list)
===================================

1. GET    /api/v1/watch-list          — 获取自选股列表
2. POST   /api/v1/watch-list          — 添加自选股
3. PUT    /api/v1/watch-list/{id}     — 更新自选股
4. DELETE /api/v1/watch-list/{id}     — 删除自选股
"""

import logging

from fastapi import APIRouter, HTTPException, Request

from api.deps import get_effective_user_uid
from api.v1.schemas.watch_list import (
    WatchListItemCreate,
    WatchListItemResponse,
    WatchListItemUpdate,
    WatchListResponse,
)
from src.repositories.watch_list_repo import WatchListRepo

logger = logging.getLogger(__name__)
router = APIRouter()


def _repo() -> WatchListRepo:
    return WatchListRepo()


@router.get("", response_model=WatchListResponse, summary="获取自选股列表")
def list_watch_list(http_request: Request):
    uid = get_effective_user_uid(http_request)
    items = _repo().list_all(uid=uid)
    return WatchListResponse(
        items=[WatchListItemResponse.model_validate(i) for i in items],
        total=len(items),
    )


@router.post("", response_model=WatchListItemResponse, status_code=201, summary="添加自选股")
def create_watch_list_item(http_request: Request, body: WatchListItemCreate):
    uid = get_effective_user_uid(http_request)
    repo = _repo()
    if repo.get_by_code(body.code, uid=uid):
        raise HTTPException(status_code=409, detail=f"股票 {body.code} 已在自选股中")
    try:
        item = repo.create(
            uid=uid,
            code=body.code,
            name=body.name,
            notes=body.notes,
            market_type=body.market_type,
            is_favorite=body.is_favorite,
        )
    except Exception as e:
        logger.error("创建自选股失败: %s", e)
        raise HTTPException(status_code=500, detail="创建失败，请重试") from e
    return WatchListItemResponse.model_validate(item)


@router.put("/{item_id}", response_model=WatchListItemResponse, summary="更新自选股")
def update_watch_list_item(http_request: Request, item_id: int, body: WatchListItemUpdate):
    uid = get_effective_user_uid(http_request)
    item = _repo().update(
        item_id=item_id,
        uid=uid,
        name=body.name,
        notes=body.notes,
        market_type=body.market_type,
        is_favorite=body.is_favorite,
    )
    if item is None:
        raise HTTPException(status_code=404, detail="未找到该自选股")
    return WatchListItemResponse.model_validate(item)


@router.delete("/{item_id}", status_code=204, summary="删除自选股")
def delete_watch_list_item(http_request: Request, item_id: int):
    uid = get_effective_user_uid(http_request)
    deleted = _repo().delete(item_id, uid=uid)
    if not deleted:
        raise HTTPException(status_code=404, detail="未找到该自选股")
